src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -20,6 +20,158 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
                 TakeIterator<TSource>(source, count);
         }
 
+        /// <summary>Returns a specified range of contiguous elements from a sequence.</summary>
+        /// <param name="source">The sequence to return elements from.</param>
+        /// <param name="range">The range of elements to return, which has start and end indexes either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        /// <returns>An <see cref="IEnumerable{T}" /> that contains the specified <paramref name="range" /> of elements from the <paramref name="source" /> sequence.</returns>
+        public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> source, Range range)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            Index start = range.Start;
+            Index end = range.End;
+            bool isStartIndexFromEnd = start.IsFromEnd;
+            bool isEndIndexFromEnd = end.IsFromEnd;
+            int startIndex = start.Value;
+            int endIndex = end.Value;
+            Debug.Assert(startIndex >= 0);
+            Debug.Assert(endIndex >= 0);
+
+            if (isStartIndexFromEnd)
+            {
+                if (startIndex == 0 || (isEndIndexFromEnd && endIndex >= startIndex))
+                {
+                    return Empty<TSource>();
+                }
+            }
+            else if (!isEndIndexFromEnd)
+            {
+                return startIndex >= endIndex
+                    ? Empty<TSource>()
+                    : source.Skip(startIndex).Take(endIndex - startIndex);
```

Comment:
In the future, I think it might make more sense for the old `Skip` and `Take` methods to be implemented on top of `Take(Range)` rather than the opposite. Do you think it might make sense to implement the logic from scratch? (and avoid layering two enumerators in the process)

Reply:
This utilizes the internal partitioner. I was using the following code:
```cs
if (!isStartIndexFromEnd && !isEndIndexFromEnd)
{
    return startIndexValue >= endIndexValue
        ? Empty<TSource>()
        : source switch
        {
            IPartition<TSource> partition => partition.Skip(startIndexValue).Take(endIndexValue - startIndexValue),
            IList<TSource> list => new ListPartition<TSource>(list, startIndexValue, endIndexValue - 1),
            _ => new EnumerablePartition<TSource>(source, startIndexValue, endIndexValue - 1)
        };
}
```
Then I changed it to `source.Skip(startIndex).Take(endIndex - startIndex)`, because
@stephentoub :
> This is going to break once #47918 goes in.

Reply:
Interesting. @stephentoub do you think it might make sense to move the non-fromEnd `TakeIterator(Range)` implementation into  separate `.SizeOpt` and `.SpeedOpt` implementations? This is currently layering two enumerators for the most common use case.

Reply:
No strong opinion.  If you think it's worthwhile to avoid the one extra partitioner allocation, sure (note I'm drawing a distinction between partitioner and enumerator).